pkg/parquetquery/iters.go

```diff
@@ -602,11 +604,14 @@ func (t RowNumber) Preceding() RowNumber {
 // IteratorResult is a row of data with a row number and named columns of data.
 // Internally it has an unstructured list for efficient collection. The ToMap()
 // function can be used to make inspection easier.
+// WARNING: Be careful when adding slices of values from iterators to Entries, as
+// there is always the risk of the slice being reused by the iterator. Always use
+// Append or AppendValue to add entries.
 type IteratorResult struct {
 	RowNumber RowNumber
 	Entries   []struct {
-		Key   string
-		Value pq.Value
+		Key    string
+		Values []pq.Value
```

Comment:
_Starting a new thread for visibility._

My main concern is the current approach forces all parquet interaction to incur the overhead for arrays, even if not using the generic attribute columns.  I think this is a regression that we don't want. We should find a way to ensure the overhead is only incurred when necessary.  

I have two alternatives:

1.  UnionIterator

Performance aside, I kind of disagree with the abstraction of slices at the individual column level.  The grouping of values only exists conceptually based on definition level, not physically in the column. I.e. you could create a slice of all span start times in a batch or a level up at the trace.  There is no implicit grouping in the span start time itself.

The abstraction I see is:  use a UnionIterator and definitionLevel-- to collect all of the attribute values into the slice.  And read the column normally.   Only do this for the generic attribute value columns.

I think the overhead incurred is same as this PR, and only to relevant queries.  The UnionIterator would incur the cost of building up slices of parquet.Values, and AttributeCollector converts them into Static arrays.   

1. Fork SyncIterator

If the above isn't feasible, then my next vote is to for the SyncIterator to a separate version for arrays.  Only use it for the generic attribute value columns.  This has a lot of duplication of code, but performance is critical and worth it.  There would also be some forks for pooling, because the result structs are different.

The performance of this approach could possibly be a bit better than (1), since it is bypassing the layers of functions and building the slices at a lower level.

Reply:
@electron0zero and I dug into this more offline and determined that approach (1) can be even simpler.  We don't need an additional union.  The existing join will already hand all values to the attributecollector in a single IteratorResult and the values will look this for an example string array:

  values [
      { "key", k },
      { "string", s1 },
      { "string", s2 },
      ...
  ]

We can update attributecollector to build the arrays from these values.  The overall mechanism is the same:  sync iterator gives raw values, a join iterator (instead of union) builds them into slices, and attributeCollector builds them into `traceql.Static`s.

Reply:
updated.